File: python/questions/cloneGraph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    oc = {}
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None:
            return None
        ocr = self.oc.get(node, None)
        if ocr is not None:
            return self.oc[node]
        ccn = Node(node.val)
        self.oc[node] = ccn
        for n in node.neighbors:
            ccn.neighbors.append(self.cloneGraph(n))
        return ccn

File: cpp/questions/cloneGraph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
private:
    std::unordered_map<Node*, Node*> oc;
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return nullptr;
        }
        if (oc.find(node) != oc.end()) {
            return oc[node];
        }
        Node* ccn = new Node(node->val);
        oc[node] = ccn;
        for (const auto& n: node->neighbors) {
            ccn->neighbors.emplace_back(cloneGraph(n));
        }
        return ccn;
    }
};
